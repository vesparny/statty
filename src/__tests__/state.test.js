import React from 'react'
import Enzyme, { mount } from 'enzyme'
import serializer from 'enzyme-to-json/serializer'
import { Provider, State } from '../index'
import Adapter from 'enzyme-adapter-react-16'

Enzyme.configure({ adapter: new Adapter() })

const CHANNEL = '__statty__'

expect.addSnapshotSerializer(serializer)

const increment = state => ({ count: state.count + 1 })
const getMockedContext = () => ({
  [CHANNEL]: {
    broadcast: {
      getState: () => {},
      setState: () => {},
      subscribe: () => 1,
      unsubscribe: jest.fn()
    }
  }
})

test('Connector updates global state', () => {
  const state = { count: 0 }
  const wrapper = mount(
    <Provider state={state}>
      <State
        render={(state, update) => (
          <button onClick={() => update(increment)}>{state.count}</button>
        )}
      />
    </Provider>
  )
  expect(wrapper).toMatchSnapshot(`with count = 0`)
  wrapper.find('button').simulate('click')
  expect(wrapper).toMatchSnapshot(`with count = 1`)
})

test('Connector updates local state', () => {
  const wrapper = mount(
    <Provider state={{ count: 0 }}>
      <State
        state={{ count: 10 }}
        render={(state, update) => (
          <button onClick={() => update(increment)}>{state.count}</button>
        )}
      />
    </Provider>
  )
  expect(wrapper).toMatchSnapshot(`with count = 10`)
  wrapper.find('button').simulate('click')
  expect(wrapper).toMatchSnapshot(`with count = 11`)
})

test('Connector returns a slice of the state based on a selector function', () => {
  const selector = state => ({ data: state.data })
  const wrapper = mount(
    <Provider state={{ count: 0, data: [] }}>
      <State
        select={selector}
        render={(state, update) => <span>{JSON.stringify(state)}</span>}
      />
    </Provider>
  )
  expect(wrapper).toMatchSnapshot(`with selector`)
})

test('Connector returns null in case no render prop is provided', () => {
  const wrapper = mount(
    <Provider state={{ count: 0 }}>
      <State />
    </Provider>
  )
  expect(wrapper).toMatchSnapshot(`with no render prop`)
})

test('unsubscribes from state updates on unmount', () => {
  const context = getMockedContext()
  const wrapper = mount(<State render={(state, update) => <span />} />, {
    context
  })
  wrapper.unmount()
  expect(context[CHANNEL].broadcast.unsubscribe).toHaveBeenCalled()
})
