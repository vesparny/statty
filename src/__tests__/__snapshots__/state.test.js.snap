// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with count = 0 1`] = `
<Provider
  state={
    Object {
      "count": 0,
    }
  }
>
  <State
    render={[Function]}
    select={[Function]}
  >
    <button
      onClick={[Function]}
    >
      0
    </button>
  </State>
</Provider>
`;

exports[`with count = 1 1`] = `
<Provider
  state={
    Object {
      "count": 0,
    }
  }
>
  <State
    render={[Function]}
    select={[Function]}
  >
    <button
      onClick={[Function]}
    >
      1
    </button>
  </State>
</Provider>
`;

exports[`with count = 10 1`] = `
<Provider
  state={
    Object {
      "count": 0,
    }
  }
>
  <State
    render={[Function]}
    select={[Function]}
    state={
      Object {
        "count": 10,
      }
    }
  >
    <button
      onClick={[Function]}
    >
      10
    </button>
  </State>
</Provider>
`;

exports[`with count = 11 1`] = `
<Provider
  state={
    Object {
      "count": 0,
    }
  }
>
  <State
    render={[Function]}
    select={[Function]}
    state={
      Object {
        "count": 10,
      }
    }
  >
    <button
      onClick={[Function]}
    >
      11
    </button>
  </State>
</Provider>
`;

exports[`with no render prop 1`] = `
<Provider
  state={
    Object {
      "count": 0,
    }
  }
>
  <State
    render={[Function]}
    select={[Function]}
  />
</Provider>
`;

exports[`with selector 1`] = `
<Provider
  state={
    Object {
      "count": 0,
      "data": Array [],
    }
  }
>
  <State
    render={[Function]}
    select={[Function]}
  >
    <span>
      {"data":[]}
    </span>
  </State>
</Provider>
`;
